### PR TITLE
Refactors build_features to improve performance

### PR DESCRIPTION
**Pull request by cp2boston**
### Pull Request Description
This PR contains refactors within the `build_features` code path that are focused on improving the performance.  There are some new files:
- Dockerfile.arm is included for reference.  It was used to create a testing/debugging environment on a Mac M1
- .dockerignore filters out unwanted context
- tests/profiling/dfs_profile.py is a helper module to run cProfile against DFS

### CODE REVIEW

File: Dockerfile.arm [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# Dockerfile for creating an image for testing python versions
```
Comment (gsheni) [inline Dockerfile.arm:1]: Can we put this in the docs (https://featuretools.alteryx.com/en/stable/install.html) or somewhere else? If possible, I would like to keep the root directory clean and minimal.

File: featuretools/tests/profiling/dfs_profile.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  - output currently goes to the root directory and is in dfs_profile.stats
+  - *.stats is gitignored
+  - it uses the demo customers dataset for testing
+  - max_depth > 2 is very slow (currently)
+  - stats output can be viewed online with https://nejc.saje.info/pstats-viewer.html
+"""
+import cProfile
+from pathlib import Path
+
+import featuretools as ft
```
Comment (sbadithe) [inline featuretools/tests/profiling/dfs_profile.py:18]: Just out of curiosity: when is it better to use `import featuretools as ft ` v.s. `from featuretools import ... `? The tests in `featuretools/tests` mostly use `from ... import ...`. Is there any one in particular that should be favored going forward?

File: docs/source/release_notes.rst [hunk 3]
```diff
@@ -6,15 +6,18 @@ Release Notes
...
     * Fixes
     * Changes
+        * Add Dockerfile.arm for Mac M1 testing/debugging (:pr:`2224`)
+        * Add tests/profiling/dfs_profile.py (:pr:`2224`)
     * Documentation Changes
     * Testing Changes
         * Update create feedstock PR forked repo to use (:pr:`2223`)
 
     Thanks to the following people for contributing to this release:
     :user:`gsheni`
```
Comment (sbadithe) [inline docs/source/release_notes.rst:19]:
```suggestion
    :user:`cp2boston`, :user:`gsheni`
```

File: featuretools/tests/profiling/dfs_profile.py (hunk 2 above)
Comment (cp2boston) [reply featuretools/tests/profiling/dfs_profile.py:18]: tbh, I was just following the pattern in the example docs.  There's no particular benefit other than keeping things in separate namespaces

File: Dockerfile.arm (hunk 1 above)
Comment (cp2boston) [reply Dockerfile.arm:1]: How about a `docker` or `dockerfiles` subdirectory?  Or, I can remove it and update gitignore to not include it

File: docs/source/release_notes.rst (hunk 3 above)
Comment (cp2boston) [reply docs/source/release_notes.rst:19]: +1

File: Dockerfile.arm (hunk 1 above)
Comment (cp2boston) [reply Dockerfile.arm]: I removed it and will keep it locally for my own use

Comment (cp2boston) [reply Dockerfile.arm]: I removed it and will keep it locally for my own use

Comment (gsheni) [reply Dockerfile.arm]: That works.

Another option is to put it in [contributing.md](https://github.com/alteryx/featuretools/blob/main/contributing.md).

File: docs/source/release_notes.rst [hunk 4]
```diff
@@ -6,15 +6,17 @@ Release Notes
 Future Release
 ==============
     * Enhancements
+        * Refactor build_features code path for performance (:pr:`2224`)
     * Fixes
     * Changes
+        * Add tests/profiling/dfs_profile.py (:pr:`2224`)
```
Comment (thehomebrewnerd) [inline docs/source/release_notes.rst:12]: This should probably be in the testing changes section.

File: featuretools/feature_base/feature_base.py [hunk 5]
```diff
@@ -14,8 +18,12 @@
 
 _ES_REF = {}
 
+# custom caching containers
+
```
Comment (thehomebrewnerd) [inline featuretools/feature_base/feature_base.py:22]: Remove?

File: featuretools/feature_base/feature_base.py [hunk 6]
```diff
@@ -141,35 +149,46 @@ def get_dependencies(self, deep=False, ignored=None, copy=True):
 
 
         """
-        deps = []
+        hash_key = hash(f"{self.get_name()}{self.dataframe_name}{deep}{ignored}")
+        if cached_dependencies := self.cache.get(CacheType.DEPENDENCY, hash_key):
```
Comment (thehomebrewnerd) [inline featuretools/feature_base/feature_base.py:153]: A walrus in the wild - don't see many of those in our codebase!

File: featuretools/feature_base/feature_base.py [hunk 7]
```diff
@@ -141,35 +149,46 @@ def get_dependencies(self, deep=False, ignored=None, copy=True):
...
+            return cached_dependencies
 
-        for d in self.base_features[:]:
-            deps += [d]
+        ignored = ignored if ignored else set()
```
Comment (thehomebrewnerd) [inline featuretools/feature_base/feature_base.py:156]: I think you could also do `ignored = ignored or set()` here if you wanted.

Comment (cp2boston) [reply featuretools/feature_base/feature_base.py:156]: Good call!

File: featuretools/feature_base/feature_base.py (hunk 6 above)
Comment (cp2boston) [reply featuretools/feature_base/feature_base.py:153]: ikr?  If cleans things up rather nicely, i think

File: docs/source/release_notes.rst (hunk 4 above)
Comment (cp2boston) [reply docs/source/release_notes.rst:12]: will do

File: featuretools/feature_base/feature_base.py (hunk 5 above)
Comment (cp2boston) [reply featuretools/feature_base/feature_base.py:22]: Good catch!

**Comment by gsheni** [comment]
@cp2boston The tests mostly pass now but some lines are not being hit. Also there is a conflict

File: featuretools/synthesis/deep_feature_synthesis.py [hunk 8]
```diff
@@ -965,6 +995,7 @@ def _get_matching_inputs(
         primitive_options,
         require_direct_input=False,
         feature_filter=None,
+        transform_stacking=False,
```
Comment (rwedge) [inline featuretools/synthesis/deep_feature_synthesis.py:998]: `transform_stacking` as parameter name is confusing.  Currnetly, if `transform_stacking` is `True`, features that stack on top of transform features are not allowed, and when `transform_stacking` is `False`, stacking on transform feature is allowed.

Also, sinces `check_transform_stacking` excludes any combination of inputs where any input is a transform feature or a direct feature of a transform feature, I think we can filter out those two groups of features as valid input types, period, before creating input combinations (either in `_features_by_type` or by making `check_transform_stacking` a `feature_filter` for `_get_matching_inputs` when called from `_build_transform_features`

Comment (cp2boston) [reply featuretools/synthesis/deep_feature_synthesis.py:998]: I'm going to take the feature_filter approach and see how that plays out

File: featuretools/synthesis/deep_feature_synthesis.py [hunk 9]
```diff
@@ -840,8 +841,6 @@ def feature_filter(f):
             wheres = list(self.where_clauses[child_dataframe.ww.name])
 
             for matching_input in matching_inputs:
-                if not check_stacking(agg_prim, matching_input):
-                    continue
```
Comment (rwedge) [inline featuretools/synthesis/deep_feature_synthesis.py:844]: These lines should probably stay, it's this removal that is causing the discrepancy in number of features generated

File: featuretools/feature_base/feature_base.py (hunk 6 above)
Comment (ozzieD) [reply featuretools/feature_base/feature_base.py]: Never knew about these till now. I love it!

File: featuretools/synthesis/deep_feature_synthesis.py [hunk 10]
```diff
@@ -693,12 +700,11 @@ def _build_transform_features(
...
                 current_options,
                 require_direct_input=require_direct_input,
+                feature_filter=check_transform_stacking,
             )
 
             for matching_input in matching_inputs:
-                if all(
-                    bf.number_output_features == 1 for bf in matching_input
-                ) and check_transform_stacking(matching_input):
+                if not any(bf.number_output_features != 1 for bf in matching_input):
```
[the remark this answers is not in the document]
Comment (sbadithe) [reply featuretools/synthesis/deep_feature_synthesis.py]: I might have confused myself here but I think the two expressions are logically equivalent and therefore should always stop at the same point.

[the remark this answers is not in the document]
Comment (cp2boston) [reply featuretools/synthesis/deep_feature_synthesis.py]: any() will stop at the first match, all() has to run through all of them.  Performance-wise, it makes a difference

File: docs/source/release_notes.rst [hunk 11]
```diff
@@ -6,19 +6,19 @@ Release Notes
...
     * Documentation Changes
-        * Add line breaks between sections of IsFederalHoliday primitive docstring (:pr:`2235`)
     * Testing Changes
+        * Add tests/profiling/dfs_profile.py (:pr:`2224`)
         * Update create feedstock PR forked repo to use (:pr:`2223`, :pr:`2237`)
         * Update development requirements and use latest for documentation (:pr:`2225`)
 
     Thanks to the following people for contributing to this release:
-    :user:`gsheni`, :user:`ozzieD`, :user:`sbadithe`, :user:`tamargrey`
+    :user:`cp2boston`, :user:`gsheni`, :user:`ozzieD`, ::user:`sbadithe`, :user:`tamargrey`
```
[the remark this answers is not in the document]
Comment (cp2boston) [reply docs/source/release_notes.rst:23]: fixed

File: featuretools/synthesis/deep_feature_synthesis.py (hunk 9 above)
Comment (cp2boston) [reply featuretools/synthesis/deep_feature_synthesis.py:844]: Ah, good catch!

File: featuretools/synthesis/deep_feature_synthesis.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (sbadithe) [reply featuretools/synthesis/deep_feature_synthesis.py]: Oops, I was under the impression `all` stopped at the first `False`

File: docs/source/release_notes.rst [hunk 12]
```diff
@@ -6,19 +6,19 @@ Release Notes
 Future Release
 ==============
     * Enhancements
+        * Refactor build_features code path for performance (:pr:`2224`)
     * Fixes
-        * Allow boolean columns to be included in remove_highly_correlated_features (:pr:`2231`)
```
[the remark this answers is not in the document]
Comment (cp2boston) [reply docs/source/release_notes.rst:10]: I think merge from main got confused.  Added it back in

File: docs/source/release_notes.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (cp2boston) [reply docs/source/release_notes.rst:23]: Fixed

File: featuretools/synthesis/deep_feature_synthesis.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (cp2boston) [reply featuretools/synthesis/deep_feature_synthesis.py]: They are essentially opposites and your assumption is correct.  all() stops at the first false.  For whatever reason, the profile numbers improved when I switched to any()

[the remark this answers is not in the document]
Comment (sbadithe) [reply featuretools/synthesis/deep_feature_synthesis.py]: Wow, that is fascinating, and not at all what I would have expected! Interesting, thanks!

File: docs/source/release_notes.rst [hunk 13]
```diff
@@ -6,15 +6,18 @@ Release Notes
...
         * Fix compatibility issues with holidays 0.15 (:pr:`2254`)
     * Changes
         * Update release notes to make clear conda release portion (:pr:`2249`)
     * Documentation Changes
     * Testing Changes
+        * Add tests/profiling/dfs_profile.py (:pr:`2224`)
 
     Thanks to the following people for contributing to this release:
-    :user:`cp2boston`, :user:`gsheni`
+    :user:`cpark`, :user:`gsheni`
```
Comment (thehomebrewnerd) [inline docs/source/release_notes.rst:19]: Correct user name?
image

**Review by thehomebrewnerd — APPROVED** [review]
Seems fine to me, but not sure I completely followed all the refactors. Might be good to get a second opinion before merging.

**Review by rwedge — APPROVED** [review]

**Review by ozzieD — APPROVED** [review]

File: docs/source/release_notes.rst (hunk 13 above)
Comment (gsheni) [inline docs/source/release_notes.rst:19]:
```suggestion
    :user:`cp2boston`, :user:`gsheni`
```

File: docs/source/release_notes.rst [hunk 14]
```diff
@@ -6,16 +6,19 @@ Release Notes
 Future Release
 ==============
     * Enhancements
+        * Refactor build_features code path for performance (:pr:`2224`)
```
Comment (gsheni) [inline docs/source/release_notes.rst:9]:
```suggestion
        * Refactor build_features to speed up long running DFS calls by 50% (:pr:`2224`)
```

**Pull request merged**